### PR TITLE
Added tokenManager.on/off/clear and autoRefresh

### PR DESCRIPTION
**Pull request by lboyette-okta**
Resolves: OKTA-94717 and OKTA-94716

@rchild-okta 
@manueltanzi-okta

### CODE REVIEW

File: lib/TokenManager.js [hunk 1]
```diff
@@ -13,7 +14,60 @@ var cookieStorage = {
...
+  delete tokenMgmtRef.refreshTimeouts[key];
+}
+
+function clearRefreshTimeoutAll(tokenMgmtRef) {
+  var refreshTimeouts = tokenMgmtRef.refreshTimeouts;
+  for(var key in refreshTimeouts) {
+    if (!refreshTimeouts.hasOwnProperty(key)) {
+      continue;
+    }
+    clearTimeout(refreshTimeouts[key]);
```
Comment (rchild-okta) [inline lib/TokenManager.js]: nit: can we use the clearRefeshTimeout function here so we don't have this logic in two places?

File: lib/TokenManager.js [hunk 2]
```diff
@@ -13,7 +14,60 @@ var cookieStorage = {
...
+  }
+  tokenMgmtRef.refreshTimeouts = {};
+}
+
+function setRefreshTimeout(sdk, tokenMgmtRef, storage, key, token) {
+  // Wait a minimum of 0 milliseconds (the token is already expired)
```
Comment (rchild-okta) [inline lib/TokenManager.js]: nit: would be clearer to say "(if < 0, the token is already expired)"

File: lib/TokenManager.js [hunk 3]
```diff
@@ -13,7 +14,60 @@ var cookieStorage = {
...
+
+  for(var key in tokenStorage) {
+    if (!tokenStorage.hasOwnProperty(key)) {
+      continue;
+    }
+    var token = tokenStorage[key];
+    if (tokenMgmtRef.autoRefresh) {
+      setRefreshTimeout(sdk, tokenMgmtRef, storage, key, token);
+    } else {
+      tokenMgmtRef.emitter.emit('expired', key, token);
```
Comment (rchild-okta) [inline lib/TokenManager.js]: Can you explain this line? And is this supposed to be what happens:
- Get tokens and store them in the TokenManager
- Refresh the page and re-initialize tokenManager with autoRefresh = false
- All my tokens emit an "expired" message

File: lib/TokenManager.js (hunk 2 above)
Comment (lboyette-okta) [inline lib/TokenManager.js]: will fix

File: lib/TokenManager.js (hunk 1 above)
Comment (lboyette-okta) [inline lib/TokenManager.js]: we can. It adds an unnecessary `delete`, but it is clearer.

File: lib/TokenManager.js (hunk 3 above)
Comment (lboyette-okta) [inline lib/TokenManager.js]: That's the correct flow, minus the last bullet. We should only emit "expired" when we're past the token's expiration. Will fix.

File: lib/TokenManager.js [hunk 4]
```diff
@@ -13,7 +14,63 @@ var cookieStorage = {
...
+  }
+
+  for(var key in tokenStorage) {
+    if (!tokenStorage.hasOwnProperty(key)) {
+      continue;
+    }
+    var token = tokenStorage[key];
+    if (tokenMgmtRef.autoRefresh) {
+      setRefreshTimeout(sdk, tokenMgmtRef, storage, key, token);
+    } else if (token.expiresAt * 1000 < Date.now()) {
```
Comment (rchild-okta) [inline lib/TokenManager.js]: We might want to deal with this in another PR, but what happens in the case of no autoRefresh and the token hasn't expired on init? It seems like when that token expires, it won't emit an 'expired' message.

File: lib/TokenManager.js [hunk 5]
```diff
@@ -23,20 +80,27 @@ function add(storage, key, token) {
   }
   tokenStorage[key] = token;
   storage.setTokenStorage(tokenStorage);
+  if (tokenMgmtRef.autoRefresh) {
+    setRefreshTimeout(sdk, tokenMgmtRef, storage, key, token);
```
Comment (rchild-okta) [inline lib/TokenManager.js]: Also here - it seems like we might want another method that schedules emitting that event if we're not autoRefreshing.

File: lib/TokenManager.js (hunk 4 above)
Comment (lboyette-okta) [inline lib/TokenManager.js]: You're right. I'll open a Jira to track this

**Comment by rchild-okta** [comment]
Overall lgtm 🚀 !

File: lib/TokenManager.js (hunk 5 above)
Comment (lboyette-okta) [inline lib/TokenManager.js]: That's a good point. a setExpiredTimeout perhaps.

**Pull request merged**